src/main/java/org/dependencytrack/tasks/OSVDownloadTask.java

```diff
@@ -0,0 +1,284 @@
+    public Severity calculateOSVSeverity(OSVAdvisory advisory) {
+
+        // derive from database_specific cvss v3 vector if available
+        if(advisory.getCvssV3Vector() != null) {
+            Cvss cvss = Cvss.fromVector(advisory.getCvssV3Vector());
+            Score score = cvss.calculateScore();
+            return normalizedCvssV3Score(score.getBaseScore());
+        }
+        // derive from database_specific cvss v2 vector if available
+        if (advisory.getCvssV2Vector() != null) {
+            Cvss cvss = Cvss.fromVector(advisory.getCvssV2Vector());
+            Score score = cvss.calculateScore();
+            return normalizedCvssV2Score(score.getBaseScore());
+        }
+        // get database_specific severity string if available
+        if (advisory.getSeverity() != null) {
+            if (advisory.getSeverity().equalsIgnoreCase("CRITICAL")) {
+                return Severity.CRITICAL;
+            } else if (advisory.getSeverity().equalsIgnoreCase("HIGH")) {
+                return Severity.HIGH;
+            } else if (advisory.getSeverity().equalsIgnoreCase("MODERATE")) {
+                return Severity.MEDIUM;
+            } else if (advisory.getSeverity().equalsIgnoreCase("LOW")) {
+                return Severity.LOW;
+            }
+        }
+        // get largest ecosystem_specific severity from its affected packages
+        if (advisory.getVulnerabilities() != null) {
+            List<Integer> severityLevels = new ArrayList<>();
+            for (OSVVulnerability vuln : advisory.getVulnerabilities()) {
+                severityLevels.add(vuln.getSeverity().getLevel());
+            }
+            Collections.sort(severityLevels);
+            Collections.reverse(severityLevels);
+            return getSeverityByLevel(severityLevels.get(0));
+        }
+        return Severity.UNASSIGNED;
+    }
+
+    public Vulnerability.Source extractSource(String vulnId) {
+
+        final String sourceId = vulnId.split("-")[0];
+        switch (sourceId) {
+            case "GHSA": return Vulnerability.Source.GITHUB;
+            case "CVE": return Vulnerability.Source.NVD;
+            default: return Vulnerability.Source.GOOGLE;
+        }
+    }
+
+    public VulnerableSoftware mapVulnerabilityToVulnerableSoftware(final QueryManager qm, final OSVVulnerability vuln) {
+
+        String versionStartIncluding = vuln.getLowerVersionRange();
+        String versionEndExcluding = vuln.getUpperVersionRangeExcluding();
+        String versionEndIncluding = vuln.getUpperVersionRangeIncluding();
+
+        try {
+            PackageURL purl = new PackageURL(vuln.getPurl());
+            VulnerableSoftware vs = qm.getVulnerableSoftwareByPurl(purl.getType(), purl.getNamespace(), purl.getName(),
+                    versionEndExcluding, versionEndIncluding, null, versionStartIncluding);
+            if (vs != null) {
+                return vs;
+            }
+            if (vuln.getPurl() == null) return null;
```

Comment:
This `null` check should happen before `new PackageURL` is called. ATM all advisories without PURL cause an error to be logged.

```
2022-07-01 22:19:37,499 ERROR [OSVDownloadTask] Error while mapping vulnerable software null
com.github.packageurl.MalformedPackageURLException: Invalid purl: Contains an empty or null value
	at com.github.packageurl.PackageURL.parse(PackageURL.java:478)
	at com.github.packageurl.PackageURL.<init>(PackageURL.java:68)
	at org.dependencytrack.tasks.OSVDownloadTask.mapVulnerabilityToVulnerableSoftware(OSVDownloadTask.java:237)
	at org.dependencytrack.tasks.OSVDownloadTask.updateDatasource(OSVDownloadTask.java:125)
	at org.dependencytrack.tasks.OSVDownloadTask.unzipFolder(OSVDownloadTask.java:99)
	at org.dependencytrack.tasks.OSVDownloadTask.inform(OSVDownloadTask.java:72)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:99)
```